runtime/main/src/main/java/org/qbicc/runtime/main/Once.java

```diff
@@ -0,0 +1,60 @@
+package org.qbicc.runtime.main;
+
+/**
+ * A one-time operation.
+ */
+public final class Once {
+
+    private final Runnable target;
+    private final Object lock = new Object();
+    private Throwable thrown;
+    private volatile boolean done;
+
+    /**
+     * Construct a new instance.
+     *
+     * @param target the target to run
+     */
+    public Once(Runnable target) {
+        this.target = target;
+    }
+
+    /**
+     * Execute the one-time task.
+     * If it throws an exception, the thrown object propagates to the caller.
+     * If the task was previously run successfully, this method does nothing.
+     * If the task was previously run but failed, the same thrown object is propagated to the caller.
+     * If the task is in the process of being run by the current thread, this method returns without doing anything.
+     * If the task is in the process of being run by some other thread, the calling thread will block until the task completes.
+     *
+     * @throws Throwable the object thrown by the target task
+     */
+    public void run() throws Throwable {
+        boolean done = this.done;
+        if (! done) {
+            if (Thread.holdsLock(lock)) {
+                // we're already running it
+                return;
+            }
+            synchronized (lock) {
+                done = this.done;
+                if (! done) try {
+                    Runnable target = this.target;
+                    if (target != null) {
+                        target.run();
+                    }
```

Comment:
Do we want to enforce that `target` is never null?

Reply:
I wasn't sure, but I guess it's a reasonable default which can be revisited later.